brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java

```diff
@@ -0,0 +1,187 @@
+   */
+  @Override public SdkHttpRequest modifyHttpRequest(
+      Context.ModifyHttpRequest context,
+      ExecutionAttributes executionAttributes
+  ) {
+    TraceContext maybeDeferredRootSpan = executionAttributes.getAttribute(DEFERRED_ROOT_CONTEXT);
+    Span applicationSpan;
+    if (maybeDeferredRootSpan != null) {
+      Boolean sampled = httpTracing.clientSampler().trySample(
+          ADAPTER,
+          context.httpRequest().toBuilder()
+      );
+      if (sampled == null) {
+        sampled = httpTracing.tracing().sampler().isSampled(maybeDeferredRootSpan.traceId());
+      }
+      applicationSpan = tracer.toSpan(maybeDeferredRootSpan.toBuilder().sampled(sampled).build());
+      executionAttributes.putAttribute(APPLICATION_SPAN, applicationSpan.start());
+    } else {
+      applicationSpan = executionAttributes.getAttribute(APPLICATION_SPAN);
+    }
+    String serviceName = executionAttributes.getAttribute(SERVICE_NAME);
+    applicationSpan.name(serviceName);
+
+    return context.httpRequest().copy(builder -> {
+      Span clientSpan = nextClientSpan(applicationSpan, serviceName, context.request(), builder);
+      executionAttributes.putAttribute(CLIENT_SPAN, clientSpan);
+    });
+  }
+
+  /**
+   * After individual http response, may run multiple times if retries occur
+   */
+  @Override
+  public void beforeUnmarshalling(
+      Context.BeforeUnmarshalling context,
+      ExecutionAttributes executionAttributes
+  ) {
+    Span clientSpan = executionAttributes.getAttribute(CLIENT_SPAN);
+    if (!context.httpResponse().isSuccessful()) {
+      clientSpan.tag("error", context.httpResponse().statusText()
+          .orElse("Unknown AWS service error"));
+    }
+    handler.handleReceive(context.httpResponse(), null, clientSpan);
+    executionAttributes.putAttribute(CLIENT_SPAN, null);
+  }
+
+
+
+  /**
+   * After a SDK request has been executed
+   */
+  @Override public void afterExecution(
+      Context.AfterExecution context,
+      ExecutionAttributes executionAttributes
+  ) {
+    Span applicationSpan = executionAttributes.getAttribute(APPLICATION_SPAN);
+    applicationSpan.finish();
+  }
+
+  /**
+   * After a SDK request has failed
+   */
+  @Override public void onExecutionFailure(
+      Context.FailedExecution context,
+      ExecutionAttributes executionAttributes
+  ) {
+    Span applicationSpan = executionAttributes.getAttribute(APPLICATION_SPAN);
+    applicationSpan.error(context.exception());
+  }
+
+  private Span nextClientSpan(
+      Span applicationSpan,
+      String serviceName,
+      SdkRequest sdkRequest,
+      SdkHttpRequest.Builder sdkHttpRequestBuilder
+  ) {
+    Span span = tracer.newChild(applicationSpan.context());
+    // Set up http, this starts the span
+    handler.handleSend(injector, sdkHttpRequestBuilder, span);
+    // Don't start the span here
+    String operation = getAwsOperationNameFromRequestClass(sdkRequest);
```

Comment:
In an undetermined future release, the AWS SDK will also include the operation name, so this magic is hopefully temporary.

Reply:
Awesome. Good to know!